index.js

```diff
@@ -1,35 +1,23 @@
 // Public intefrace
 module.exports = promisedPipe
 
+const ensureFunctionType = (fn, i) => {
+    if (typeof fn !== 'function') {
+        throw Error(`pipe requires each argument to be a function. Argument #${i+1} is of type "${typeof fn}"`)
+    }
+}
+
+const chain = (q, fn) => q.then(fn)
+
 function promisedPipe(...fns) {
     if (fns.length < 1) {
         throw Error('pipe requires at least one argument')
     }
 
-    fns.forEach((fn, i) => {
-        if (typeof fn !== 'function') {
-            throw Error(`pipe requires each argument to be a function. Argument #${i+1} is of type "${typeof fn}"`)
-        }
-    })
+    fns.forEach(ensureFunctionType)
 
```

Comment:
Ok, so I hacked a little benchmark at `benchmark.js`.

Here's [this branch](https://travis-ci.org/kirillrogovoy/promised-pipe/jobs/198238051) vs [master](https://travis-ci.org/kirillrogovoy/promised-pipe/jobs/198298137)

I also played a lot locally with your code and came to a couple conclusions:

1. Hoisting the validator doesn't actually make the pipe creation faster on modern Node versions. Not sure about the Node 4 though. Anyway, the difference is too small.
2. The simplified way you used to create the promise chain is much faster than my naive solution.

So I just propose to go back to `forEach` and remove the hoisting and use your way of doing anything else. :)